src/main.rs

```diff
@@ -65,14 +68,26 @@ fn gumpi_async(opt: Opt, config: JobConfig) -> impl Future<Item = (), Error = fa
     // It's safe to call expect here - at this point opt.jobconfig
     // is guaranteed to be a valid filepath, which is checked by
     // JobConfig::from_file in main.rs
-    let sources_dir = opt
+    let jobconfig_dir = opt
         .jobconfig
         .parent()
         .expect("Invalid jobconfig path")
         .to_owned();
     let noclean = opt.noclean;
 
-    SessionMPI::init(opt.hub, prov_filter)
+    // The initialization of the provider may take time,
+    // so check if the file exists at all in advance
+    if let Some(input) = &config.input {
+        let input_path = jobconfig_dir.join(&input.source);
+        if !input_path.is_file() {
+            return Err(format_err!(
+                "The input data, {}, doesn't exist",
+                input_path.to_string_lossy()
```

Comment:
This is the very reason I find `PathBuf` and `Path` tedious to use!

Reply:
It's even worse: https://github.com/golemfactory/gumpi/blob/master/src/session/mpi.rs#L307-L313
(if I weren't using `into_future`, the code would get even more messy, because I'd have to handle the `Result` manually, when I'm returning a `Future`)

Reply:
Haha, yeah, that's horrible!